### PR TITLE
Deprecate camel case date format

### PR DESCRIPTION
**Pull request by pgomulka**
Camel case date formats are deprecated and snake case should be used instead.

### CODE REVIEW

File: .idea/runConfigurations/Debug_Elasticsearch.xml [hunk 1]
```diff
@@ -6,6 +6,10 @@
     <option name="HOST" value="localhost" />
     <option name="PORT" value="5005" />
     <option name="AUTO_RESTART" value="true" />
+    <RunnerSettings RunnerId="Debug">
```
Comment (rjernst) [inline .idea/runConfigurations/Debug_Elasticsearch.xml:9]: This should not be changing?

File: server/src/main/java/org/elasticsearch/common/time/FormatNames.java [hunk 2]
```diff
@@ -122,7 +139,34 @@ public static boolean exist(String format) {
...
-        return format.equals(camelCaseName) || format.equals(snakeCaseName);
+        if (format.equals(camelCaseName) || format.equals(snakeCaseName)) {
+            deprecate(format);
+            return true;
+        }
+        return false;
+    }
+
+    private void deprecate(String format) {
+        if (DEPRECATED_CAMEL_CASE_NAMES.contains(format)) {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/time/FormatNames.java:161]: Shouldn't all camelCase names be deprecated, in which case we can only log the deprecated message if the first condition matches above?

File: server/src/test/java/org/elasticsearch/common/time/FormatNamesTests.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+package org.elasticsearch.common.time;
+
+import junit.framework.TestCase;
+
+public class FormatNamesTests extends TestCase {
+
+    public void testDeprecation(){
+
```
Comment (rjernst) [inline server/src/test/java/org/elasticsearch/common/time/FormatNamesTests.java:27]: missing test?

**Review by rjernst — COMMENTED** [review]
I know this is in draft, but thought I'd mention a few things I noticed.

File: .idea/runConfigurations/Debug_Elasticsearch.xml (hunk 1 above)
Comment (pgomulka) [reply .idea/runConfigurations/Debug_Elasticsearch.xml:9]: yes, that file should not be modified.

File: server/src/main/java/org/elasticsearch/common/time/FormatNames.java (hunk 2 above)
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/common/time/FormatNames.java:161]: yes, all camel case should be deprecated. It is just for some like `iso8601`, `hour`, `year` etc snake case and camel case is the same. We do not need to log the deprecation for these

File: server/src/test/java/org/elasticsearch/common/time/FormatNamesTests.java (hunk 3 above)
Comment (pgomulka) [reply server/src/test/java/org/elasticsearch/common/time/FormatNamesTests.java:27]: will remove this. I added deprecation test to https://github.com/elastic/elasticsearch/pull/58909/files#diff-198e1ff52598409599e5a427493f0227R386

File: server/src/main/java/org/elasticsearch/common/time/FormatNames.java (hunk 2 above)
Comment (rjernst) [reply server/src/main/java/org/elasticsearch/common/time/FormatNames.java:161]: In that case perhaps we could not fill in the camelCaseName and condition on that being non null? It seems like we are doing extra work to analyze the casing of the name when we have already separated the ok (snake case) from the deprecated (camelCase).

Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/common/time/FormatNames.java:161]: I see what you mean now. That's the smart way to do it - will fix!

File: server/src/main/java/org/elasticsearch/common/time/FormatNames.java [hunk 4]
```diff
@@ -19,13 +19,17 @@
...
+import org.apache.logging.log4j.LogManager;
+import org.elasticsearch.common.logging.DeprecationLogger;
+import org.elasticsearch.common.util.LazyInitializable;
+
 import java.util.Arrays;
 import java.util.Set;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 public enum FormatNames {
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/common/time/FormatNames.java:31]: for the record. this class was already used by `Joda` class. By mistake it was not used by `DateFormatters`

File: server/src/main/java/org/elasticsearch/common/time/FormatNames.java [hunk 5]
```diff
@@ -108,11 +112,19 @@
...
+
     private final String camelCaseName;
     private final String snakeCaseName;
 
+    // DateFormatters and FormatNames are being used even before the logging is initialized.
+    // If LogManager.getLogger is called before logging config is loaded
+    // it results in errors sent to status logger and startup to fail.
+    // Hence a lazy initialization.
+    private static final LazyInitializable<DeprecationLogger, RuntimeException> deprecationLogger
+        = new LazyInitializable(() -> new DeprecationLogger(LogManager.getLogger(FormatNames.class)));
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/common/time/FormatNames.java:126]: as per comment. Some classes - which are used before loading log4j config  - are using `DateFormat.forPattern` to get a date format. see `XContentElasticsearchExtension` 
This is fine, but if I try to create a logger, before logging config is loaded I will get an exception (no config, writing to status logger) this is later checked and startup will fail.
I am not sure we can move loading log4j config as early on startup. 
At the same time, this logging is a temporary. We will remove it in `8` (unless we want some more detailed exception)
@rjernst WDYT?

**Review by rjernst — CHANGES_REQUESTED** [review]
One general comment: I think this PR needs to be against master, and backported to 7.x. Otherwise we lose the history that we first deprecated, and the the camelCase names still exist in master.

**Comment by pgomulka** [comment]
> One general comment: I think this PR needs to be against master, and backported to 7.x. Otherwise we lose the history that we first deprecated, and the the camelCase names still exist in master.

@rjernst we already have a PR that forbids using camelCase names https://github.com/elastic/elasticsearch/pull/59417 I can refactor that PR (#59417 against master) to deprecate, then cherry-pick to 7.x and finally remove it in master in another PR. 
There will be some conflicts though, `FormatNames` did not exist in master before. It was meant for deprecation of joda formats in 7.x initially.
What do you mean by loosing history? We can always link these two PRs with `relates 1`

**Comment by pgomulka** [comment]
that totally makes sens to me now. thanks for clarifying. I will follow up with possible a different PR/force push on this one

**Comment by pgomulka** [comment]
closing in favour of https://github.com/elastic/elasticsearch/pull/59555 against master
will backport to v7

**Pull request closed without merging**